sources/PTYSession.m

```diff
@@ -13343,23 +13343,84 @@ - (void)screenDidUpdateReturnCodeForMark:(id<VT100ScreenMarkReading>)mark
     [self screenNeedsRedraw];
 }
 
-- (void)screenPostUserNotification:(NSString * _Nonnull)message {
+- (void)screenPostUserNotification:(NSString * _Nonnull)message rich:(BOOL)rich {
     if (![self shouldPostTerminalGeneratedAlert]) {
         DLog(@"Declining to allow terminal to post user notification %@", message);
         return;
     }
     DLog(@"Terminal posting user notification %@", message);
     [self incrementBadge];
-    NSString *description = [NSString stringWithFormat:@"Session %@ #%d: %@",
-                             [[self name] removingHTMLFromTabTitleIfNeeded],
-                             [_delegate tabNumber],
-                             message];
-    [[iTermNotificationController sharedInstance]
-     notify:@"Alert"
-     withDescription:description
-     windowIndex:[self screenWindowIndex]
-     tabIndex:[self screenTabIndex]
-     viewIndex:[self screenViewIndex]];
+
+    iTermNotificationController* controller = [iTermNotificationController sharedInstance];
+
+    if (rich) {
+        NSArray *parts = [message componentsSeparatedByString:@"|"];
+
+        if ([parts count] != 5) {
+            DLog(@"Incorrect number of parts in rich notification, %lu vs 5: %@",
+                    [parts count], message);
+            return;
+        }
+
+        NSString *title = nil;
+        NSString *subtitle = nil;
+        NSString *description = nil;
+        NSString *image = nil;
+        NSString *sound = nil;
+
+        NSMutableArray *decoded_parts = [[NSMutableArray alloc] init];
+        // Decode all the arguments
+        for (NSString* part in parts) {
+            NSString* decoded_part = [part stringByBase64DecodingStringWithEncoding:NSUTF8StringEncoding];
+            if (decoded_part == nil) {
+                DLog(@"Rich notification field not valid base64 data: %@", part);
+                return;
+            }
+            [decoded_parts addObject: decoded_part];
+        }
+
+        // Description
+        NSString* part = decoded_parts[2];
+        if ([part length] > 0) {
+            description = part;
+        } else {
+            DLog(@"Rich notification requires a non-empty description field.");
+        }
+
+        // Title
+        part = decoded_parts[0];
```

Comment:
I recommend using key-value pairs rather than assuming an order. It'll make it easier to modify this in the future. Like this:

```OSC 1337 ; Notification=title=[base64] ; subtitle=[base64] ; image=[base64] ST```

Reply:
I was trying to keep the code simpler by avoiding the need to do more complex parsing.  I've rewritten it to parse key-value pairs and it proved to not be that much more code and effort.